### ISSUE TITLE
The checkout process needs to be stable and atomic

### ISSUE THREAD

**Issue by Konamiman** [woothemes/woocommerce#43770]
### Describe the bug

We have got reports of the checkout process breaking for various reasons and the an order being created in a "half-baked" (inconsistent) state as a result:

[https://github.com/woocommerce/woocommerce/issues/14541](<https://github.com/woocommerce/woocommerce/issues/14541>)<br>[https://github.com/woocommerce/woocommerce/issues/41527](<https://github.com/woocommerce/woocommerce/issues/41527>)<br>[https://github.com/woocommerce/woocommerce/issues/43648](<https://github.com/woocommerce/woocommerce/issues/43648>)

[WOOPLUG-5341](https://linear.app/a8c/issue/WOOPLUG-5341/investigate-empty-order-addresses-on-block-checkout)

[WOOPLUG-6099](https://linear.app/a8c/issue/WOOPLUG-6099/duplicate-orders-created-from-concurrent-checkout-requests)

The "duplicate orders" behavior is a symptom of this, often caused by users retrying the submission of the order (but we don't know for sure if this is always the cause).

### Expected behavior

The checkout process needs to be atomic, this means that either the checkout succeeds (and then the order confirmation successfully reaches the customer) or it fails (and then no trace/leftovers of the order are present in the server).

**Comment by lkraav** [comment]
I feel like this touches https://github.com/woocommerce/woocommerce/issues/24603 type things, too.

**Comment by adam-sandor** [comment]
Has there been any progress on solving the broken/duplicate order problems? I see a whole plethora of Github tickets open but no meaningful progress on any of them. We are still getting broken orders because of oom error every day.

**Comment by wavvves** [comment]
Hi @Konamiman, I have been looking into https://github.com/woocommerce/woocommerce/issues/41527. While I agree with this, despite the effort to make placing an order robust and remove any doubt about success/failure, it will always depend on payment gateway confirmation, which might be outside our reach.
What I mean with this is, shopper places an order, the charge is attempted and we depend on the external service robustness to get a confirmation back. Well ideally, we could be double-checking the charge in an async way which might also fail, and on the positive outcome would definitely increase the time to process a request before redirecting to order confirmation.

Is there any plan or hypothesis to tackle this?

One way I was thinking about implementing to fight back the possibility of the shopper placing several orders, would be through feedback.
I wonder how many times this doesn't happen because the shopper gets notified of the charge on a different device (phone), although the order seemed to fail on checkout screen. If we can also output some feedback there when payments don't respond as expected, this might greatly help in reducing hassles or at least buy us some time to rethink the process payment flow.

That being said I'm planning to open a PR that would output a notice warning something has not gone as expected, and to check the orders on my account, check charges, or contact the merchant before retrying in the case we don't get a valid response back. Any thoughts on this?

**Comment by adam-sandor** [comment]
I would just like to add here that duplicate orders are not the full problem. Even when the user doesn't place a second order we have orders where the payment has succeeded but stock has not been deducted and other processes have not ran.
Also for us it's always an OOM error that causes things to fail. The PHP-FPM process gets into some kind of loop until it uses up all the available memory.

**Comment by lkraav** [comment]
> The PHP-FPM process gets into some kind of loop until it uses up all the available memory.

Y'all are running Sentry or some equivalent stack monitoring platform, right? These might be able snag a backtrace before PHP-FPM OOMs.

**Comment by adam-sandor** [comment]
That would be great but Cloudways customer support didn't offer anything when I was trying to get their help in solving this.

**Comment by lkraav** [comment]
> That would be great but Cloudways customer support didn't offer anything when I was trying to get their help in solving this.

Yep, at a certain scale, over-managed / -jailed hosting becomes a bottleneck of its own.

**Comment by wavvves** [comment]
Internal discussion: pca54o-8fk-p2
